src/error/metadata/code.rs

```diff
@@ -1,11 +1,114 @@
+    // include for the explanation md below :)
+    #[allow(clippy::upper_case_acronyms)]
+    EALL,
+}
 
 impl Display for Code {
     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(formatter, "{:?}", &self)
     }
 }
+
+impl Code {
+    // builds a BTreeMap of every possible code and its explanation, so we can
+    // access from the `explain` function and get a single one OR so we can
+    // iterate over them in `explain_all` for creating a docs page
+    fn explanations() -> BTreeMap<Code, String> {
+        let contents = vec![
+            (Code::E001, include_str!("./codes/E001.md").to_string()),
+            (Code::E002, include_str!("./codes/E002.md").to_string()),
+            (Code::E003, include_str!("./codes/E003.md").to_string()),
+            (Code::E004, include_str!("./codes/E004.md").to_string()),
+            (Code::E005, include_str!("./codes/E005.md").to_string()),
+            (Code::E006, include_str!("./codes/E006.md").to_string()),
+            (Code::E007, include_str!("./codes/E007.md").to_string()),
+            (Code::E008, include_str!("./codes/E008.md").to_string()),
+            (Code::E009, include_str!("./codes/E009.md").to_string()),
+            (Code::E010, include_str!("./codes/E010.md").to_string()),
+            (Code::E011, include_str!("./codes/E011.md").to_string()),
+            (Code::E012, include_str!("./codes/E012.md").to_string()),
+            (Code::E013, include_str!("./codes/E013.md").to_string()),
+            (Code::E014, include_str!("./codes/E014.md").to_string()),
+            (Code::E015, include_str!("./codes/E015.md").to_string()),
+            (Code::E016, include_str!("./codes/E016.md").to_string()),
+            (Code::E017, include_str!("./codes/E017.md").to_string()),
+            (Code::E018, include_str!("./codes/E018.md").to_string()),
+            (Code::E019, include_str!("./codes/E019.md").to_string()),
+            (Code::E020, include_str!("./codes/E020.md").to_string()),
+            (Code::E021, include_str!("./codes/E021.md").to_string()),
+            (Code::E022, include_str!("./codes/E022.md").to_string()),
+            (Code::E023, include_str!("./codes/E023.md").to_string()),
+            (Code::E024, include_str!("./codes/E024.md").to_string()),
+            (Code::E025, include_str!("./codes/E025.md").to_string()),
+            (Code::E026, include_str!("./codes/E026.md").to_string()),
+            (Code::E027, include_str!("./codes/E027.md").to_string()),
+        ];
+        contents.into_iter().collect()
+    }
+
+    /// For a given error code, returns a markdown string with a given error's
+    /// explanation. Explanations are in ./codes
+    pub fn explain(&self) -> String {
+        let all_explanations = Code::explanations();
+
+        match self {
+            // return all error explanations, concated with headings, for docs
+            Code::EALL => {
+                let mut all_md: String = "".to_string();
```

Comment:
It's best to use `String::new` for empty allocations
```suggestion
                let mut all_md: String = String::new();
```

Reply:
reallly???? why? i was always told it was strictly personal preference

Reply:
`String::new()` for empty string does not do the extra initial buffer allocation that a `"".to_string()` does! Super fun little tidbit! Honestly, the allocation difference is quite small especially for a CLI that it doesn't matter in this case. Might matter for things like encoders or parsers or for cases where you might never actually fill up the string o/

Technically the most efficient way in this particular case would be to do `String::with_capacity()` and calculate the allocation amount necessary for all of the md strings, but that's very tedious and will have to be changed with every error code we write, so just having a `String::new()` is ok for me.